Update doc for new LDAP configuration (#241)
exoplatform/exo-documentation: docs/LDAP.rst

Please refer to this list to check defaults values of these parameters 
and define them in the ``exo.properties`` file to adapt them to your directory characteristics.

Once the parameters are set, the eXo Platform can be started and users and groups will be imported.

.. _LDAP.ConfigurationReference:

=======================
Configuration reference
=======================

Here is the list of all available configuration properties for directory integration to define in ``exo.properties``:

   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+-----------------------------------------+
   | Name                                           | Description                                             | Value                     | Default                            | Example                                 |
   +================================================+=========================================================+===========================+====================================+=========================================+

>    | exo.ldap.type                                  | Type of LDAP server                                     | ``ldap``, ``ad`` or empty | Empty (no LDAP/AD integration)     | cn                                      |

Reviewer: A typo here I think
```suggestion
   | exo.ldap.type                                  | Type of LDAP server                                     | ``ldap``, ``ad`` or empty | Empty (no LDAP/AD integration)     | ldap                                      |
```

Author: Right, thanks. Done.

    -  :ref:`Synchronization Configuration <LDAP.SynchronizationConfiguration>`
       A guide of all available configuration parameters for directory synchronization.

    -  :ref:`Advanced configuration <LDAP.AdvancedConfiguration>`
       A guide about advanced configuration using PicketLink IDM configuration.

    -  :ref:`Frequently asked questions <LDAP.FAQ>`
       How to resolve some possible issues of a directory integration.

.. _LDAP.Introduction:

============
Introduction
============

eXo Platform can be plugged to a directory server to use its users, groups and memberships:

-  the directory can be plugged in read-only mode only

-  the directory can contain users and groups, or only users

-  the supported directory implementations are: OpenLDAP and Microsoft Active Directory. You can refer to our official 
   `supported environments <https://www.exoplatform.com/terms-conditions/supported-environments.pdf>`__ matrix for more 
   details about the supported versions. 

The term "Directory users" represents users who are created in the directory by its utilities.
The term "Platform users" represents users who are created via eXo Platform UI.

> The understanding is similar for "Directory groups" and "Platform group*".

Reviewer 2: ```suggestion
The understanding is similar for "Directory groups" and "Platform groups".
```

 -  **exo.ldap.type** - defines the type of directory. Must be ``ad`` for Microsoft Active Directory or ``ldap`` for any other LDAP directory.
 -  **exo.ldap.url** - defines the URL to connect to the directory.
 -  **exo.ldap.admin.dn** - defines the full DN of the admin user used to fetch data.
 -  **exo.ldap.admin.password** - defines the password of the admin user used to fetch data.
 -  **exo.ldap.users.base.dn** - defines the base DN of the users. Multiple DNs can be provided by separating them by semicolons.
 -  **exo.ldap.groups.base.dn** - (optional, required only if directory groups must be used) defines the base DN of the groups. Multiple DNs can be provided by separating them by semicolons.

Example:

   .. code:: properties

       exo.ldap.type=ldap
       exo.ldap.url=ldap://my-ldap-server-host:389
       exo.ldap.admin.dn=cn=admin,dc=company,dc=com
       exo.ldap.admin.password=123456
       exo.ldap.users.base.dn=ou=users1,dc=company,dc=com;ou=users2,dc=company,dc=com
       exo.ldap.groups.base.dn=ou=groups,dc=company,dc=com

The chapter `Configuration reference <LDAP.ConfigurationReference>`__ lists all of available parameters.

> Please refer to this list to check defaults values of these parameters 

Reviewer 3: ```suggestion
Please refer to this list to check default values of these parameters 
```

Author: Done

and define them in the ``exo.properties`` file to adapt them to your directory characteristics.

> Once the parameters are set, the eXo Platform can be started and users and groups will be imported.

Reviewer 3: ```suggestion
Once the parameters are set, the eXo Platform can be started. Users and groups will be then imported.
```

Author: Done

   |                                                | and removed from the pool,                              |                           |                                    |                            |
   |                                                | alias for com.sun.jndi.ldap.connect.pool.timeout        |                           |                                    |                            |
   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+----------------------------+
   | exo.ldap.connection.pool.protocol              | List of space-separated protocol types of               | ``plain`` and/or ``ssl``  | ``plain ssl``                      |                            |
   |                                                | connections that may be pooled,                         |                           |                                    |                            |

>    |                                                | alias for com.sun.jndi.ldap.connect.pool.protocol       |                           |                                    |                            |

Reviewer 2: it is not clear wether this property can be defined in the exo.properties onot what happens is tit's defined as a System property. Does exo.ldap.connection.pool.protocol overrides com.sun.jndi.ldap.connect.pool.protocol in that case or is it just ignored ?

Author: I don't know, and I don't find any doc about that, I will check in the source code and let you know.

Author: "exo.ldap.*" properties override the "com.sun.*" for [those defined in the picketlink configuration](https://github.com/Meeds-io/gatein-portal/pull/16/files#diff-6b70d8d0aaf06025d73f66871106c7f9R316-R322). After doing some tests, the behavior is quite confusing:
* there are 2 ways to define these variables
* variables overridden in picketlink configuration are set only when the first LDAP connection is initiated, which means that the value of com.sun.* properties can have a first value at startup, then another value after the first LDAP connection

I propose to remove these custom variables in PicketLink configuration. Administrators will simply customize LDAP connection pool by defining com.sun.* properties directly, either in exo.properties or in bin/setenv-customize.sh. And defaults values will be [JVM default values](https://docs.oracle.com/javase/jndi/tutorial/ldap/connect/config.html). We use the standards.
WDYT ?

Author: Done. Tell me if it is not ok for you.

   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+----------------------------+
   | exo.ldap.search.timelimit                      | Number of milliseconds that a search may last           | Any positive number       | 10000                              |                            |
   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+----------------------------+

>    | exo.ldap.users.base.dn                         | Semicolon-separated list of full DNs                    |                           | ``ou=users,dc=company,dc=org``     |                            |

Reviewer 4: can should we put an example with more than one value ?

Author: Yes, done

   | exo.ldap.users.attributes.{name}.isMultivalued | Is the attribute multi-valued                           | ``true`` or ``false``     | ``false``                          |                            |
   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+----------------------------+
   | exo.ldap.users.search.scope                    | Scope of the search for users                           | ``base``, ``one``         | ``subtree``                        |                            |
   |                                                |                                                         | or ``subtree``            |                                    |                            |
   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+----------------------------+

>    | exo.ldap.groups.base.dn                        | Semicolon-separated list of full DNs                    | ``true`` or ``false``     | ``ou=groups,dc=company,dc=org``    |                            |

Reviewer 4: should we show an example with more than one value ?

Author: Yes, done

   | exo.ldap.groups.rootGroup                      | Root group to bind LDAP/AD groups, all LDAP/AD groups   | Any group id, ending      | ``/platform/*``                    |                            |
   |                                                | will be available under this group. It must end         | with ``/``                |                                    |                            |
   |                                                | with ``/``. Root group ("/") cannot be used.            |                           |                                    |                            |
   +------------------------------------------------+---------------------------------------------------------+---------------------------+------------------------------------+----------------------------+
 
.. _LDAP.SynchronizationConfiguration:

=============================
Synchronization configuration
=============================

The LDAP integration uses jobs to synchronize periodically the eXo internal database with the data modified in the LDAP.

> The periodicy of these jobs can be changed in ``exo.properties`` thanks to the following properties:

Reviewer 3: ```suggestion
The periodicity of these jobs could be cutomized in ``exo.properties`` file thanks to the following properties:
```

Author: Done

.. code-block:: jproperties

    # Cron expression used to schedule the job that will import periodically data
    # from external store
    # (Default value = every ten minutes)
    exo.idm.externalStore.import.cronExpression=0 */10 * ? * *
    # Cron expression used to schedule the job that will delete periodically data
    # from internal store that has been deleted from external store
    # (Default value = every day at 23:59 PM)
    exo.idm.externalStore.delete.cronExpression=0 59 23 ? * *
    # Cron expression used to schedule the job that will process periodically data
    # injected in queue
    # (Default value = every minute)
    exo.idm.externalStore.queue.processing.cronExpression=0 */1 * ? * *

> By default, user data are synced with the LDAP when the user signs in. This can be disabled by modifying this property:

Reviewer 3: ```suggestion
By default, user data are synchronized with the LDAP when he signs in. This can be disabled by modifying this property:
```

Author: Done
